### ISSUE TITLE
Custom Inline Templates

### ISSUE THREAD

**Issue by mAAdhaTTah** [tinacms/tinacms#3058]
## Summary

As a developer I want to customize the inline marks of the rich text editor so that I can support footnotes & other inline elements

### Expected Results

* Developer can add inline marks to the rich text field

### Notes

My use case is footnotes but I could see this being useful for other marks, like highlights.

**Comment by jamespohalloran** [comment]
Hey! Do you mind explaining a bit more exactly what you're looking for here, (and provide a code example if possible).

**Comment by mAAdhaTTah** [comment]
Sure. So the general idea is I have some markdown with footnotes that I'm trying to migrate into Tina's rich text field. In markdown, that looks like this:

```md
This is a sentence with a footnote.[^1]

[^1]: And this is a footnote.
```

To migrate this into the rich text field, I need to add a new inline element (similar to bold, italic, & code). So the above paragraph could be represented something like this following (if I'm remember the schema correctly):

```json
{
  "type": "paragraph",
  "children": [
    {
      "type": "text",
      "text": "This is a sentence with a footnote."
    },
    {
      "type": "text",
      "text": "And this is a footnote"
      "footnote": true
    }
  ]
}
```

This is an "inline" element with a new `footnote` property. When I render said element, I can replace them with a link and render the collected footnotes at the bottom of the page. This is my specific goal.

More generally, I was looking at the rich text field an thinking that it might be more useful for the editor itself to be extensible enough to enable these inline marks to be added more generically, rather than enabling footnotes specifically, hence the suggestion to enable "inline marks" rather than "footnotes". Another usecase might be highlighting, with a `"highlight": true` property on a text node. The overall idea is to make the rich text editor more extensible, starting with inline marks.

Does that make sense?

**Comment by DmacMcgreg** [comment]
This would be super helpful.
Also supporting superscript tags would be a blessing.
Right now, I'm running my own Superscript react component to get around this.

**Comment by mAAdhaTTah** [comment]
Additional use case: There are a number of inline tags that are unsupported right now: `kbd`, `abbr`, `sup`, `sub`, `q`, etc. I don't know if it makes sense to support all of them in core but it would be great to be able to enable support per-element in userland. In order for this to work, it needs to hook into (or at least modify) both the Plate editor as well as the serialization format.

**Comment by jeffsee55** [comment]
> as well as the serialization format.

Yeah this is the part that makes granular customizations difficult. We _should_ support the marks the CommonMark and GFM support, but probably not all of them since we're handling things through MDX. footnotes and tables are good examples, but for anything that's not in these specs  our answer will be to use MDX. The problem there is that a tiny little MDX concern like being able to add a `sup` tag around a text node feels heavy-handed. You'd define the `sup` MDX template on the field, and then the editor would need to step through that into the nested form for that value. It'd be ideal to be able to define some inline MDX templates as truly inline.

So right now for `sup`, you'd define this:

```
{
  type: 'rich-text',
  name: 'body',
  templates: [{
    name: 'sup',
    label: 'Superscript',
    fields: [{
      type: 'string',
      label: 'Value'
      name: 'value',
      inline: true,
      isTitle: true // note this was recently added (it shows this value)
    }]
  }]
}
```
And that looks like this:

<img width="413" alt="Screen Shot 2022-08-30 at 2 46 02 PM" src="https://user-images.githubusercontent.com/5414297/187549349-e65f48ef-5701-4126-8fea-e183d2500cfd.png">

But for these scenarios the field which is `isTitle` should probably be editable from within the main view without needing to step through. It's something we can probably do, but the UX needs some thought.

**Comment by gaurav712** [comment]
@jeffsee55 The `inline: true` doesn't work for me. I have the exact same useCase. An extract from the schema:
```
const superscriptTemplate = {
  label: "SuperScript",
  name: "superScript",
  fields: [
    {
      type: "string",
      label: "Value",
      name: "value",
      inline: true,
    },
  ],
};

const chapterCollection = {
  label: "Chapter",
  name: "chapterEntry",
  fields: [
    {
      type: "string",
      name: "chapterName",
      label: "Chapter Name",
    },
    {
      type: "boolean",
      name: "hasAR",
      label: "Has AR?",
    },
    {
      type: "string",
      name: "modelName",
      label: "AR Model File Name",
    },
    {
      type: "string",
      component: "select",
      name: "theme",
      label: "Button Theme",
      description: "Select a theme for the button",
      options: ["Dark", "Light"],
    },
    {
      type: "rich-text",
      name: "content",
      label: "Content",
      isBody: true,
      templates: [
        headingTemplate,
        subHeadingTemplate,
        imageTemplate,
        videoTemplate,
        arPlayerTemplate,
        superscriptTemplate,
      ],
    },
  ],
};

```
Am I missing anything?

**Comment by jeffsee55** [comment]
@gaurav712 that should be:
```
const superscriptTemplate = {
  label: "SuperScript",
  name: "superScript",
  inline: true,
  fields: [
    {
      type: "string",
      label: "Value",
      name: "value",
    },
  ],
};

```

**Comment by gaurav712** [comment]
@jeffsee55 Cheers! Worked.

**Comment by mAAdhaTTah** [comment]
Oh... I didn't realize the `templates` could be `inline`. That might solve a lot of those use cases without getting so deep into customizing the editor.

**Comment by mAAdhaTTah** [comment]
Ok, so coming back to this after working through some other issues I ran into (mostly around content migration). I have a footnote as follows:

```md
This is a sentence with a footnote.[^1]

[^1]: This is a footnote with a [link](https://jamesdigioia.com).
```

I defined by Footnote as follows:

```js
const footnoteTemplate = {
  name: 'Footnote',
  label: 'Footnote',
  inline: true,
  fields: [
    {
      name: 'children',
      label: 'Note',
      type: 'rich-text',
    },
  ],
};
```

If I type the above into the editor, the mdx saved by TinaCMS looks like this:

```mdx
This is a sentence with a footnote.<Footnote>This is a footnote with a [link](https://jamesdigioia.com).</Footnote>
```

But when I reload the editor and navigate into the Footnote, I get this error:

```
Error parsing markdown
Content: text is not yet supported at line: 1, column: 36

To fix these errors, edit your content locally and then restart the Tina server.
```

If I were to change it to be as follows:

```js
/**
 * @type RichTextField['template']
 */
const footnoteTemplate = {
  name: 'Footnote',
  label: 'Footnote',
  inline: true,
  fields: [
    {
      name: 'note',
      label: 'Note',
      type: 'rich-text',
    },
  ],
};
```

It's saved this way:

```mdx
This is a sentence with a footnote.<Footnote note={<>This is a footnote with a [link](https://jamesdigioia.com).</>} />
```

But instead of an error when I navigate to the footnote, the `note` field is empty.

On some level, maybe this is a bit of a corner case (how often do inline elements need extra block elements associated with them?), but perhaps the thing to be fixed is Tina lets you set this up this way then doesn't behave correctly once it is, rather than telling you it's not possible. Also means there's a case for integrating footnotes natively, rather than through MDX.

**Comment by jeffsee55** [comment]
not stale

**Comment by jeffsee55** [comment]
not stale

**Comment by DmacMcgreg** [comment]
Not stale

**Comment by DmacMcgreg** [comment]
Bumping this, since we're doing some awful workarounds for this issue...

**Comment by jaschaio** [comment]
@DmacMcgreg whats your workaround?

**Comment by jaschaio** [comment]
not stale

**Comment by kulesy** [comment]
Closing as stale — this enhancement request is from July 2022 (3.5 years old). The template system has changed significantly since. Part of backlog refinement.
